jormungandr/src/explorer/mod.rs

```diff
@@ -66,112 +89,303 @@ impl Explorer {
 
-    pub fn get_header(
+    pub fn get_block(
         &self,
-        hash: HeaderHash,
-    ) -> impl Future<Item = Option<Header>, Error = Infallible> {
-        //XXX: Probably the clone is not necessary
-        self.multiverse
-            .get(hash)
-            .map(|maybe_block_ref| maybe_block_ref.map(|r| (*r.header()).clone()))
+        block_id: &HeaderHash,
+    ) -> impl Future<Item = Option<ExplorerBlock>, Error = Infallible> {
+        let multiverse = self.multiverse.clone();
+        let block_id = block_id.clone();
+        get_lock(&self.longest_chain_tip).and_then(move |tip| {
+            multiverse.get((*tip).id()).and_then(move |maybe_state| {
+                let state = maybe_state.expect("the longest chain to be indexed");
+                Ok(state.blocks.lookup(&block_id).map(|b| (*b).clone()))
+            })
+        })
     }
 
     pub fn find_block_by_transaction(
         &self,
-        transaction: FragmentId,
-    ) -> impl Future<Item = Option<HeaderHash>, Error = StorageError> {
-        let mut blocks = self.transaction_to_block.clone();
-        future::poll_fn(move || Ok(blocks.poll_lock()))
-            .and_then(move |guard| Ok(guard.get(&transaction).map(|h| (*h).clone())))
+        transaction: &FragmentId,
+    ) -> impl Future<Item = Option<HeaderHash>, Error = Infallible> {
+        let multiverse = self.multiverse.clone();
+        let transaction = transaction.clone();
+        get_lock(&self.longest_chain_tip).and_then(move |tip| {
+            multiverse.get((*tip).id()).and_then(move |maybe_state| {
+                let state = maybe_state.expect("the longest chain to be indexed");
+                Ok(state.transactions.lookup(&transaction).map(|id| id.clone()))
+            })
+        })
+    }
+}
+
+fn get_lock<L>(lock: &Lock<L>) -> impl Future<Item = LockGuard<L>, Error = Infallible> {
+    let mut lock = (*lock).clone();
+    future::poll_fn(move || Ok(lock.poll_lock()))
+}
+
+fn apply_block_to_transactions(transactions: Transactions, block: &Block) -> Result<Transactions> {
+    let block_id = block.id();
+    let ids = block
+        .contents
+        .iter()
+        .filter(|fragment| is_transaction(fragment))
+        .map(|fragment| fragment.id());
+
+    let mut transactions = transactions;
+    for id in ids {
+        transactions = transactions
+            .insert(id, block_id)
+            .map_err(|_| ErrorKind::TransactionAlreadyExists(format!("{}", id)))?;
+    }
+
+    Ok(transactions)
+}
+
+fn apply_block_to_blocks(blocks: Blocks, block: &Block) -> Result<Blocks> {
+    let block_id = block.id();
+    blocks
+        .insert(block_id, (*block).clone())
+        .map_err(|_| Error::from(ErrorKind::BlockAlreadyExists(format!("{}", block_id))))
+}
+
+fn apply_block_to_addresses(addresses: Addresses, block: &Block) -> Result<Addresses> {
+    // FIXME: Probably there is a cleaner way to do this
+    let txs = block.contents.iter().filter_map(|fragment| match fragment {
```

Comment:
you are ignoring the certificate contents here. I'd have this pattern match in the loop below and do the right thing depending of the kind of fragment it is

Reply:
You will also need to split this function into small ones to help have something more _precise_.

Reply:
for sure. I need to improve that piece of code.